### PR TITLE
Dynamic translation of call, video, mute, unmute

### LINKED ISSUE #2259: Tooltips on mute microphone etc... aren't translated on language change when in a call but are when muted/unmuted
How to reproduce:
- Start a call
- Change language
- look at tooltip

### PR DESCRIPTION
**Pull request by TheNain38**
fixes #2259
Could someone check if this is right?

### CODE REVIEW

File: src/widget/form/chatform.cpp [hunk 1]
```diff
@@ -1085,5 +1085,40 @@ void ChatForm::hideNetcam()
 
 void ChatForm::retranslateUi()
 {
+    QString callObjectName = callButton->objectName();
+    QString videoObjectName = videoButton->objectName();
+    QString volObjectName = volButton->objectName();
+    QString micObjectName = micButton->objectName();
     loadHistoryAction->setText(tr("Load chat history..."));
+
+    if (callObjectName == "green")
```
Comment (antis81) [inline src/widget/form/chatform.cpp]: Better use `QStringLiteral("green")` here (and for the following conditions).

File: src/widget/form/chatform.cpp [hunk 2]
```diff
@@ -1085,5 +1085,40 @@ void ChatForm::hideNetcam()
...
+        callButton->setToolTip(tr("Start audio call"));
+
+    if (callObjectName == "yellow")
```
Comment (antis81) [inline src/widget/form/chatform.cpp]: `else if` :)

File: src/widget/form/chatform.cpp [hunk 3]
```diff
@@ -1085,5 +1085,40 @@ void ChatForm::hideNetcam()
...
+        callButton->setToolTip(tr("Accept audio call"));
+
+    if (callObjectName == "red")
+        callButton->setToolTip(tr("End audio call"));
+
+
+    if (videoObjectName == "green")
+        videoButton->setToolTip(tr("Start video call"));
+
+    if (videoObjectName == "yellow")
```
Comment (antis81) [inline src/widget/form/chatform.cpp]: `else if`…

Comment (agilob) [inline src/widget/form/chatform.cpp]: Yup, elseif of switch would work better here.

File: src/widget/form/chatform.cpp [hunk 4]
```diff
@@ -1085,5 +1085,47 @@ void ChatForm::hideNetcam()
 
 void ChatForm::retranslateUi()
 {
+    QString volObjectName = volButton->objectName();
+    QString micObjectName = micButton->objectName();
     loadHistoryAction->setText(tr("Load chat history..."));
+
+    switch(callButton->objectName())
+    {
+    case QStringLiteral("green"):
```
Comment (TheNain38) [inline src/widget/form/chatform.cpp]: I can't use QStringLiteral in `case QStringLiteral("green"):` ?

Comment (antis81) [inline src/widget/form/chatform.cpp]: That comes natural to switch-case statement. It doesn't work with `"literals"` (or other dynamic types) -> must be enumeration type (int, enum, …).

File: src/widget/form/chatform.cpp [hunk 5]
```diff
@@ -1085,5 +1085,47 @@ void ChatForm::hideNetcam()
...
+        videoButton->setToolTip(tr("Start video call"));
+        break;
+
+    case QStringLiteral("yellow"):
+        videoButton->setToolTip(tr("Accept video call"));
+        break;
+
+    case QStringLiteral("red"):
+        videoButton->setToolTip(tr("End video call"));
```
Comment (antis81) [inline src/widget/form/chatform.cpp]: Just a general tip, when using switch-case statements:
Don't forget to `break;` in **every** case. And usually you want to define a `default: break;` as well.

File: src/widget/form/chatform.cpp (hunk 4 above)
Comment (TheNain38) [inline src/widget/form/chatform.cpp]: @antis81 How did I do that correctly, then? with a switch?

File: src/widget/form/chatform.cpp (hunk 5 above)
Comment (TheNain38) [inline src/widget/form/chatform.cpp]: Yes I forgot the `default:` one

**Comment by TheNain38** [comment]
good to merge

File: src/widget/form/chatform.cpp (hunk 4 above)
Comment (antis81) [inline src/widget/form/chatform.cpp]: Short answer: You can't -> use `if (); else if(); else;`.

Long answer:
Switching the `objectName` to indicate the status is **bullshit**. We need to query the actual status of the underlying object. That means:

``` c++
// this is some pseudo code, just to give an idea
// tox API should send some kind of event (callback), if an audio-/video call starts/mutes/ends/…
// We react to the status change via an event handler and set the tooltip text according to the status:
void GenericChatForm::callStatusChanged(Call::Type t, Call::Status s)
{
    QString description;

    // assuming c++11 "enum class Type {Audio, Video, …}" here
    if (t == Call::Type::Audio) {
        switch (s) {
        default: break;
        case Call::Status::Active: description = tr("Hangup and end the current call."); break;
        case Call::Status::Ended: description = tr("Start a new  call."); break;
        case Call::Status::Pending: description = tr("Accept the call.!"); break;
        }
        callButton->setToolTip(description);
    }
    else if (t == Call::Type::Video) {
        switch (s) {
        default: break;
        case Call::Status::Active: description = tr("End the current video call."); break;
        case Call::Status::Ended: description = tr("Start a new video call."); break;
        case Call::Status::Pending: description = tr("Accept the video call.!"); break;
        }
        videoButton->setToolTip(description);
    } else {
            // undefined call type -> must not happen
            Q_ASSERT(false);
    }

    // undefined call status -> must not happen
    Q_ASSERT(!description.isEmpty());
}


```

File: src/widget/form/chatform.cpp [hunk 6]
```diff
@@ -1085,5 +1085,33 @@ void ChatForm::hideNetcam()
...
+    else if (callObjectName == QStringLiteral("red"))
+        callButton->setToolTip(tr("End audio call"));
+
+    if (videoObjectName == QStringLiteral("green"))
+        videoButton->setToolTip(tr("Start video call"));
+    else if (videoObjectName == QStringLiteral("yellow"))
+        videoButton->setToolTip(tr("Accept video call"));
+    else if (videoObjectName == QStringLiteral("red"))
+        videoButton->setToolTip(tr("End video call"));
+
```
Comment (antis81) [inline src/widget/form/chatform.cpp]: Don't you think, that moving the two if-blocks for `callObjectName` and  `videoObjectName` to the base class is the better approach?

Comment (TheNain38) [inline src/widget/form/chatform.cpp]: @antis81 Moving them under line 1116?

Comment (antis81) [inline src/widget/form/chatform.cpp]: No, I mean moving them to "genericchatform.cpp from line 520 downwards" :)

Comment (TheNain38) [inline src/widget/form/chatform.cpp]: @antis81 Ok, when done it can be merged?

**Comment by TheNain38** [comment]
@antis81 done, ready for merge

**Pull request merged**